Add TranspileLayout struct to rust

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds a Rust TranspileLayout object which is analgous to the Python space object. It is a self contained struct that is only for Rust and will be used by the C API for the C transpiler so that users will be able to reason about the permutations caused by the transpiler. This will also be used by #14106 to apply a transpilation layout to the SparseObservable in C. This commit only adds the rust struct and it's associated tests, a subsequent PR will add the C API on top. This won't be used until we have a full path transpiler in C though (see #14760).

### Details and comments

mtreinish committed:
Add TranspileLayout struct to rust

This commit adds a Rust TranspileLayout object which is analgous to the
Python space object. It is a self contained struct that is only for
Rust and will be used by the C API for the C transpiler so that users
will be able to reason about the permutations caused by the transpiler.
This will also be used by #14106 to apply a transpilation layout to the
SparseObservable in C. This commit only adds the rust struct and it's
associated tests, a subsequent PR will add the C API on top. This won't
be used until we have a full path transpiler in C though (see #14760).

jakelishman reviewed (commented):
If we're going to do this, please can we replace the Python-space version at the same time? `TranspileLayout` in Python is already not well documented and hard to follow[^1], and I really don't want _two_ parallel structs of the same type in the library at the same time.

[^1]: as this notification popped, I was _literally_ writing something against `TranspileLayout` and had to got to the `PassManager` to figure out what one of the input fields was supposed to represent.

jakelishman commented on crates/transpiler/src/transpile_layout.rs:
```diff
@@ -0,0 +1,493 @@
+    routing_permutation: Option<Vec<PhysicalQubit>>,
+    input_qubit_count: u32,
+    output_qubit_count: u32,
+}
```
This could do with more documentation, especially `routing_permutation` - this needs to clearly state how the permutation of `routing_permutation` is supposed to be interpreted.

Under what circumstances can `output_qubit_count` be different to the length of `initial_layout`?

mtreinish replied:
I added more details to the docs in: https://github.com/Qiskit/qiskit/pull/14778/commits/fbb85c9cff57c54e3098279848a3bba28a6c103d

jakelishman commented on crates/transpiler/src/transpile_layout.rs:
```diff
@@ -0,0 +1,493 @@
+                .map(PhysicalQubit::new)
+                .collect(),
+        }
+    }
```
If this is supposed to be a Rust-native struct, this would be better returning `Option<&[PhysicalQubit]>` rather than eagerly allocating.

Something like an `fn explicit_routing_permutation` could return `Cow<'self, [PhysicalQubit]>` and be essentially this method, but only allocate for the `None` case.

mtreinish replied:
I cleaned this up in: https://github.com/Qiskit/qiskit/pull/14778/commits/ba662c443bfd8cc546c9dedbfb99644d2c39c59d

jakelishman commented on crates/transpiler/src/transpile_layout.rs:
```diff
@@ -0,0 +1,493 @@
+                qubit_idx
+            })
+            .collect()
+    }
```
Please let's not make the same mistake as the Python struct, and instead return a clean `NLayout` that corresponds to the Python struct's `TranspileLayout.final_virtual_layout` / `TranspileLayout.final_index_layout` with the `VirtualQubit`s being the incoming qubits.  I think that's the same as the vector that comes out of this, but it's not possible to misinterpret because of the typing.

mtreinish replied:
I didn't use an nlayout for this because I was thinking from the c API to just return a pointer to the array of u32s. Nlayout makes that a bit harder to work with. The vec here is analogous to `final_index_layout()` in python. 

Or were you saying make this two methods, one for nlayout one for the vec?

mtreinish replied:
I made 2 methods here one for an NLayout and the other for the vec return in: https://github.com/Qiskit/qiskit/pull/14778/commits/464dcca24f6cc8967145e5c31865dc1d9c384cd3

mtreinish commented:
> If we're going to do this, please can we replace the Python-space version at the same time? `TranspileLayout` in Python is already not well documented and hard to follow[1](#user-content-fn-1-5de68aa40edff11d83cbc5463197f44f), and I really don't want _two_ parallel structs of the same type in the library at the same time.
> ## Footnotes
> 
>     1. as this notification popped, I was _literally_ writing something against `TranspileLayout` and had to got to the `PassManager` to figure out what one of the input fields was supposed to represent. [↩](#user-content-fnref-1-5de68aa40edff11d83cbc5463197f44f)

I considered doing it but I wrote this specifically using the Rust native types to work with the transpiler in rust and c. If you look at #14760 besides the output I was planning to use this as the container throughout the `PassManager` equivalent to work with all the layouts which we can't do in Python either.

The problem with using it as a replacement in python is the existing class has public attributes that are all python types and the construction is in terms of the python `Layout`. Also the passmanagers work explicitly with the layout class in different property set fields, we can't change those types or supersede those fields without breaking custom pipelines. So the best we'd really be able to do is take the `Layout` objects from the property set construct the rust types from the `Layout` objects to create the rust transpile layout object. Then we'd have to wrap this in a python specific layer that will convert the rust types to the python types when needed (and probably also storing the extra information about shareable qubits we don't need to store in rust). It didn't seem like that buys us very much.

mtreinish committed:
Add routing permutation composition method

mtreinish committed:
Improve routing permutation APIs and make more rust native

This commit improves the routing permutation method to make them more
rust native. It updates the return to be a slice of PhysicalQubits
instead of an owned vec. To get the explicit trivial permutation in the
case of None a new method explicit_routing_permutation is added that
returns `Cow<[PhysicalQubits]>` and only allocates in the None case.

mtreinish committed:
Update final layout methods

This commit updates the final layout method to return a NLayout and a
new method final_index_layout is added to return a Vec.

mtreinish committed:
Merge branch 'main' into transpile-layout-rs

mtreinish committed:
Merge remote-tracking branch 'origin/main' into transpile-layout-rs

mtreinish committed:
Improve documentation

mtreinish committed:
Merge remote-tracking branch 'origin/main' into transpile-layout-rs

mtreinish committed:
Fix lint

mtreinish committed:
Merge remote-tracking branch 'origin/main' into transpile-layout-rs

mtreinish committed:
Add support for python interop

This commit adds methods to convert to and from the Python space
transpile layout class. This requires changing the data in the rust
struct slightly so that we are tracking the virtual bits to build the
view that the Python struct needs. This isn't strictly needed for a pure
rust view of the data, but doesn't add much overhead to track it to
support python.

mtreinish committed:
Fix lint

alexanderivrii commented on crates/transpiler/src/transpile_layout.rs:
```diff
@@ -0,0 +1,1041 @@
+pub struct TranspileLayout {
+    /// The initial layout which is mapping the virtual qubits in the input circuit to the
+    /// transpiler to the physical qubits used on the transpilation target
+    initial_layout: NLayout,
```
Based on a comment I've seen somewhere, do we want to make `initial_layout` into an `Option`? In particular, we would then be able to easily determine whether the layout is already applied.

Terminology-wise, would it make sense to call this `layout` rather than `initial_layout`?

mtreinish replied:
Sure I think making this optional makes sense, as for the naming I know that you and @jakelishman have a plan on refactoring the way we reason about layout in the transpiler. I went with initial layout because that's what it was called in Python, but I'm fine calling it `layout` if that's preferable I don't really have much preference as long as it's clear to everyone.

jakelishman replied:
For naming: imo `initial_layout` is fine.  Since our plan is to get rid of `final_layout` as something that's _explicitly_ stored, I agree that `initial_` isn't needed for disambiguation, but in the prospective new object, there's still going to be "initial relabelling" and "final implicit [stuff]", so "initial" is still fine, and makes things clear.

mtreinish replied:
I made it optional in: https://github.com/Qiskit/qiskit/pull/14778/commits/83b4ee1824442ac29f99a905f3bc935cc78f379f but kept the name as `initial_layout`.

alexanderivrii commented on crates/transpiler/src/transpile_layout.rs:
```diff
@@ -0,0 +1,1041 @@
+    /// represents the permutation caused by routing or permutation elision during
+    /// transpilation. This vector maps the qubits at the start of the circuit to their
+    /// final position/physical qubit at the end of the circuit.
+    routing_permutation: Option<Vec<PhysicalQubit>>,
```
Suppose we start with a virtual circuit and run the `ElidePermutations` pass, producing a permutation of virtual qubits. Right now we store this permutation as the `"virtual_permutation_layout"` property of the property set, but in the future we want this to be stored inside this rust-space `TranspileLayout`, correct? Would it be better to use the same struct member `routing_permutation` to represent a permutation of either the virtual or the physical qubits, or would it be better to save both in this struct?

Terminology-wise, I like `"routing_permutation"` infinitely more than `"final_layout"` that we have in the python class, yet the word `"routing"` confuses me as it seems to imply that this is somehow associated with a routing pass (and not say `ElidePermutations` pass). Can we find an even better name?

mtreinish replied:
I think we should save both, I'll add the extra. How about two properties `virtual_permutation` and `physical_permutation`? Or `virtual_output_permutation` and `physical_output_permutation`?

jakelishman replied:
Why should we save both?  We already overwrite (combine) virtual permutations if you say run `ElidePermutations` and `StarPreRouting` together, and it is fundamentally exactly the same thing as a permutation induced by routing.  The only difference is which side of "apply layout" we happened to run it on.

I think keeping them separate makes it harder to interpret for both the transpiler and users than unifying them would.

jakelishman replied:
Fwiw, my PR #14826 outlines the general principles Sasha and I were talking about in the comments of the `TranspileLayout.from_property_set` method, including how "virtual permutation" and "routing permutation" are handled/unified throughout the entire pipeline.

mtreinish replied:
I assumed you wanted to track them separately and combine them in the output methods. But from #14826 I see the intent better. I'll leave it as one, how about I rename it `output_permutation` instead then?

mtreinish replied:
I went with `output_permutation` in: https://github.com/Qiskit/qiskit/pull/14778/commits/40246fbaa5ef2269e48e38f537cec0c9fb0fa8d7 if there is a better name suggested we can change it again.

alexanderivrii replied:
It's still of type ``Option<Vec<PhysicalQubit>>``, right? How can we use this type to represent the permutation of virtual qubits (after eliding permutations) before the layout takes place?

mtreinish replied:
I changed the type in: https://github.com/Qiskit/qiskit/pull/14778/commits/2d420ee31cf4dc473ea72aa827f48da3207f6dfe

alexanderivrii reviewed (commented):
I very much like how this PR looks like. As a start, I have left two preliminary questions.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into transpile-layout-rs

mtreinish committed:
Make initial_layout optional

This commit updates the initial_layout attribute to make it optional. In
the transpiler a layout doesn't necessarily need to be set so to account
for that this the attribute is made optional.

mtreinish committed:
Rename routing_permutation -> output_permutation

mtreinish committed:
Make output_permutation use Qubit instead of PhysicalQubit

Now that output_permutation is storing the abstract permutation whether
there is a layout or not it means the PhysicalQubit type doesn't
accurately reflect the purpose of the field anymore since it might not
be a physical qubit. This commit fixes this by using the Qubit type
which just represent the index in the circuit which is all the
permutation is.

jakelishman committed:
Fix Python accesses in `TranspileLayout::from_py_native`

jakelishman reviewed (approved):
I think the members of the `struct` in this object are sensible now, at least for the state of what current Qiskit tracks additionally to the DAG during transpilation.

I've been deliberately using this in the `ApplyLayout` port to try and keep things future-looking there and to get some experience using this. I think there's quite a lot we could do to improve the Rust-space access to this object, especially _during_ the transpile when we need to update it, not just query it.  I want to merge now to unblock other PRs, though, since we're free to completely change the public interface to this over time, since it's just a Rust-only object.
